tests/remotes/webdav.py

```diff
@@ -17,19 +16,27 @@
 
 
 class Webdav(Base, WebDAVURLInfo):
+    _DIR_PATH = None
+
     @staticmethod
     def get_url(port):  # pylint: disable=arguments-differ
         return f"webdav://localhost:{port}"
 
     def mkdir(self, mode=0o777, parents=False, exist_ok=False):
-        assert parents
-        makedirs(self.path, exist_ok=exist_ok)
+        self.dir_path.mkdir(parents=parents, exist_ok=True)
```

Comment:
@efiop this is because, when you are generating a directory with 2 files in it (`{'data': {'foo': 'foo', 'bar': 'bar'}}`) it tries to recreate the parent for 2 times, which results with a `DirectoryExists` error. 
https://github.com/iterative/dvc/blob/fc77618ad1eb9d0bb306a85e4f750660b5285318/tests/remotes/base.py#L43

Apparently all other trees have some sort of `assert not exist_ok` in their `mkdir()` implementation, so instead of passing this in the `Base.gen()` I just fixed here.
https://github.com/iterative/dvc/blob/fc77618ad1eb9d0bb306a85e4f750660b5285318/tests/remotes/s3.py#L77
https://github.com/iterative/dvc/blob/fc77618ad1eb9d0bb306a85e4f750660b5285318/tests/remotes/ssh.py#L84
etc.

Reply:
So now `exist_ok` is just ignored, which is not nice :( We should go the other way around and pass exist_ok=True in `_gen`.

Reply:
Those asserts in s3 and others are just defensive ones and could be changed accordingly.

Reply:
> Those asserts in s3 and others are just defensive ones and could be changed accordingly.

So should I remove all of them?

Reply:
@isidentical Only where it makes sense.

Reply:
Well, this is going to be passed to all of the trees (wherever you use `.gen()`)

Reply:
@isidentical I understand. It does make sense for gen to pass `exist_ok=True` to makedirs, as you've explained above :pray: . So it makes to adjust the incorrect/improper asserts in particular makedirs implementations. E.g. for `s3` there are no dirs at all and that assert is just defensive to make sure we are not doing anything weird, which means that we are totally fine removing it.